src/libraries/System.Net.Sockets/tests/FunctionalTests/ReceiveMessageFrom.cs

```diff
@@ -115,25 +115,42 @@ public async Task ClosedDuringOperation_Throws_ObjectDisposedExceptionOrSocketEx
                 return;
             }
 
-            using var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
-            socket.BindToAnonymousPort(IPAddress.Any);
-
-            Task receiveTask = ReceiveMessageFromAsync(socket, new byte[1], GetGetDummyTestEndpoint());
-            await Task.Delay(100);
-            if (closeOrDispose) socket.Close();
-            else socket.Dispose();
-
-            if (UsesApm)
+            int msDelay = 100;
+            if (UsesSync)
             {
-                await Assert.ThrowsAsync<ObjectDisposedException>(() => receiveTask)
-                    .TimeoutAfter(CancellationTestTimeout);
+                // In sync case Dispose may happen before the operation is started,
+                // in that case we would see an ObjectDisposedException instead of a SocketException.
+                // We may need to try the run a couple of times to deal with the timing race.
+                await RetryHelper.ExecuteAsync(() => RunTestAsync(), maxAttempts: 10, retryWhen: e => e is XunitException);
             }
             else
             {
-                SocketException ex = await Assert.ThrowsAsync<SocketException>(() => receiveTask)
-                    .TimeoutAfter(CancellationTestTimeout);
-                SocketError expectedError = UsesSync ? SocketError.Interrupted : SocketError.OperationAborted;
-                Assert.Equal(expectedError, ex.SocketErrorCode);
+                await RunTestAsync();
+            }
+
+            async Task RunTestAsync()
+            {
+                using var socket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
+                socket.BindToAnonymousPort(IPAddress.Any);
+
+                Task receiveTask = ReceiveMessageFromAsync(socket, new byte[1], GetGetDummyTestEndpoint());
+                await Task.Delay(msDelay);
+                msDelay *= 2;
```

Comment:
why the msDelay chages hare. I don't see it used laster. Also the seems generally problematic as the time needed can vary greatly across platforms. Would it work if we simply accept either exception if we feel that should be expected?

Reply:
> I don't see it used later. 

The test is being retried 10 times in the sync case, doubling the delay time on each failure:
https://github.com/dotnet/runtime/blob/7adf3db24bab6fe6ed8055aa9513f654cd3168b7/src/libraries/System.Net.Sockets/tests/FunctionalTests/ReceiveFrom.cs#L118

> the time needed can vary greatly across platforms

The retry mechanism is here to address this. Most of the times the test will succeed within the 100ms delay, on slow platforms / high load first attempts will fail with an ObjectDisposedException, but eventually one of the subsequent retries will observe a `SocketException`

I adapted this practice from an existing test introduced in dotnet/corefx#38804:

https://github.com/dotnet/runtime/blob/7adf3db24bab6fe6ed8055aa9513f654cd3168b7/src/libraries/System.Net.Sockets/tests/FunctionalTests/SendReceive/SendReceive.cs#L938-L943